### PR TITLE
Add course table, add course_id column to assignment

### LINKED ISSUE #960: Extend nbgrader.db to cope with multiple courses (and think beyond that)
The current `nbgrader.db` database has no concept of a course: each course has to have it's own database (configured in the config file)

In our environment, we have

- multiple instructors on a course, and 
- one instructor doing multiple courses

We also want to consider how to move to an external database (postgresql/mysql/etc).... but we are also very conscious that the notebook-servers themselves should *not* have access to create databases in our server (that could be horribly abused :chuckle: )

Our initial thought is to add a `courses` table, and link assignments to that

Keen to hear who else has been down this route, and what solutions have been tried.

### PR DESCRIPTION
**Pull request by BertR**
This will add a course table and a foreign key from the assignment table into the course table.

Closes #960

### CODE REVIEW

**Comment by jhamrick** [comment]
Hey @BertR , just wanted to see how this was coming along? This is the last PR needed for the multiple classes support 🙂 

Just as a heads up I'm going to be away beginning on Sunday for two weeks and will have minimal time to look at stuff then, in case you want to get any feedback from me on it before then!

**Comment by BertR** [comment]
HI @jhamrick, this now passes the test and should work as a first stepping stone towards
having configurable courses.

File: nbgrader/api.py [hunk 1]
```diff
@@ -1375,34 +1384,36 @@ def set_course_id(self, course_id, **kwargs):
...
-    def course_id(self):
-        try:
-            course_id_record = self.db.query(Course)\
-                .one()
-            return course_id_record.id
-        except NoResultFound:
-            raise MissingEntry("No course id")
+        return course
+
+    # @property
```
Comment (jhamrick) [inline nbgrader/api.py]: Could you remove this commented code?

File: nbgrader/alembic/versions/e43177bfe90b_added_course_table_and_relationships.py [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+    if len(results) == 0:
+        connection.execute(
+            "INSERT INTO course (id) VALUES ('{}')".format(default_course))
+
+    with op.batch_alter_table("assignment") as batch_op:
+
+        batch_op.add_column(sa.Column(
+            'course_id', sa.VARCHAR(128), nullable=False, default=default_course))
+        batch_op.create_foreign_key(
```
Comment (jhamrick) [inline nbgrader/alembic/versions/e43177bfe90b_added_course_table_and_relationships.py]: Do you know if `fk_course_assignment` is the same name as the foreign key created when the database is created from scratch? If not I'm just wondering if that could potentially cause problems later down the line...

**Comment by jhamrick** [comment]
Thinking about it a bit more actually, I guess it won't work properly with multiple courses yet since it would actually require changing all the database queries to specify the course id. I think that's ok for now: we can make the other changes in another PR. For now, if you just address the more minor comments that I left, I'm happy to merge this 🙂

File: nbgrader/api.py (hunk 1 above)
Comment (jhamrick) [reply nbgrader/api.py]: Looks like the commented code is still there, unfortunately...

File: nbgrader/tests/api/test_gradebook.py [hunk 3]
```diff
@@ -433,6 +433,33 @@ def test_remove_notebook(assignment):
             assignment.find_notebook(nb.name, 'foo')
 
 
+def test_course_id_constructor():
```
Comment (jhamrick) [inline nbgrader/tests/api/test_gradebook.py]: Thanks for adding these tests!

File: nbgrader/alembic/versions/e43177bfe90b_added_course_table_and_relationships.py (hunk 2 above)
Comment (BertR) [reply nbgrader/alembic/versions/e43177bfe90b_added_course_table_and_relationships.py]: In Sqlite, the foreign key constraint is not named. The migration seems to work in both directions, up and down, but I don't know whether I've considered all cases.

Comment (jhamrick) [reply nbgrader/alembic/versions/e43177bfe90b_added_course_table_and_relationships.py]: Ok, sounds good! Thanks for clarifying. I think it's fine if the constraint normally doesn't have an name.

File: nbgrader/alembic/versions/e43177bfe90b_added_course_table_and_relationships.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+def upgrade():
+    """
+    This migrations adds a course column to the assignment table
+    and a matching foreign key into the course table
+    """
+
+    # If the course table is empty, create a default course
+    connection = op.get_bind()
+    res = connection.execute("select id from course")
```
Comment (jhamrick) [inline nbgrader/alembic/versions/e43177bfe90b_added_course_table_and_relationships.py]: This line will fail if the `course` table does not yet exist.

**Comment by jhamrick** [comment]
Ok, #1147 is now merged so if you could rebase (or merge) that would be great!

**Pull request merged**